### PR TITLE
RDBS-11 Retrieving ids for a patch by index modified to use information from Lucene without accessing documents table

### PR DESCRIPTION
**Pull request by haludi**
https://issues.hibernatingrhinos.com/issue/RDBS-11

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 1]
```diff
@@ -2288,168 +2288,239 @@ public IndexDefinition GetIndexDefinition()
...
+                }
+
+                return;
+            }
+        }
 
-                                var totalResults = new Reference<int>();
-                                var skippedResults = new Reference<int>();
-                                IncludeCountersCommand includeCountersCommand = null;
+        protected class QueryContext : IDisposable
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs:2380]: Can you make this a struct? Will this cause allocations?

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 2]
```diff
@@ -2288,168 +2288,239 @@ public IndexDefinition GetIndexDefinition()
...
+            public TransactionOperationContext Context;
+            public bool IsStale;
+            public IDisposable ReturnContextToPool { get; set; }
 
-                                var fieldsToFetch = new FieldsToFetch(query, Definition);
+            public void Dispose()
+            {
+                using (ReturnContextToPool)
+                {
+                    Context.CloseTransaction();
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs:2390]: IIRC, doesn't this already happen?

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 3]
```diff
@@ -2288,168 +2288,239 @@ public IndexDefinition GetIndexDefinition()
...
+                    GetOrAddSpatialField,
+                    token.Token);
+            }
 
-                                var retriever = GetQueryResultRetriever(query, queryScope, documentsContext, fieldsToFetch, includeDocumentsCommand);
+            return documents;
+        }
 
-                                IEnumerable<(Document Result, Dictionary<string, Dictionary<string, string[]>> Highlightings, ExplanationResult Explanation)> documents;
+        protected async Task<QueryContext> OpenTransactionsForNonStaleResultIfNeeded(
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs:2460]: Can we use a ValueTask here?

File: src/Raven.Server/Documents/Queries/DocumentQueryResult.cs [hunk 4]
```diff
@@ -80,4 +80,44 @@ private static string CreateExplanation(Lucene.Net.Search.Explanation explanatio
             return explanation.ToString();
         }
     }
+
+    public class IdsQueryResult : QueryResultServerSide<string>
+    {
+        public Queue<string> Ids { get; } = new Queue<string>();
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/DocumentQueryResult.cs:86]: Why a queue and not a list?

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 2 above)
Comment (haludi) [reply src/Raven.Server/Documents/Indexes/Index.cs:2390]: As it was implemented so far the read transaction is closed after we retrieve all the ids
I just refactoring this so I can reuse the same code in query for documents and in query for ids

File: src/Raven.Server/Documents/Queries/DocumentQueryResult.cs (hunk 4 above)
Comment (haludi) [reply src/Raven.Server/Documents/Queries/DocumentQueryResult.cs:86]: `ExecuteRateLimitedOperations` using a queue

**Comment by ppekrol** [comment]
We will take a different approach on the subject. Please open a new PR when ready.

**Pull request closed without merging**